// react-persian-calendar-date-picker/lib/shared/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DAY_SHAPE = exports.WEEK_DAYS = exports.PERSIAN_MONTHS = exports.PERSIAN_NUMBERS = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var PERSIAN_NUMBERS = ['۰', '۱', '۲', '۳', '۴', '۵', '۶', '۷', '۸', '۹'];
exports.PERSIAN_NUMBERS = PERSIAN_NUMBERS;
var PERSIAN_MONTHS = ['فروردین', 'اردیبهشت', 'خرداد', 'تیر', 'مرداد', 'شهریور', 'مهر', 'آبان', 'آذر', 'دی', 'بهمن', 'اسفند'];
exports.PERSIAN_MONTHS = PERSIAN_MONTHS;
var WEEK_DAYS = {
  saturday: 'شنبه',
  sunday: 'یکشنبه',
  monday: 'دوشنبه',
  tuesday: 'سه شنبه',
  wednesday: 'چهارشنبه',
  thursday: 'پنجشنبه',
  friday: 'جمعه'
};
exports.WEEK_DAYS = WEEK_DAYS;
var DAY_SHAPE = {
  year: _propTypes["default"].number.isRequired,
  month: _propTypes["default"].number.isRequired,
  day: _propTypes["default"].number.isRequired
};
exports.DAY_SHAPE = DAY_SHAPE;
